ansible/library/bgp_route.py

```diff
@@ -0,0 +1,343 @@
+            if state == HEADER:
+                if regex_prefix_header.match(line) and (prefix not in line):
+                    self.facts['bgp_route'][prefix]['found'] = False
+                    return
+                if regex_prefix_header.match(line) and prefix in line:
+                    self.facts['bgp_route'][prefix]['found'] = True
+                    state = PATH_PROPERTIES
+                else:
+                    state = ERR
+            elif state == PATH_PROPERTIES:
+                find_path = regex_prefix_avail_paths.match(line)
+                if find_path:
+                    path_num = find_path.group(1)
+                    self.facts['bgp_route'][prefix]['path_num'] = path_num
+                    state = PEER_GROUP_HEADER
+                else:
+                    state = ERR
+            elif  state == PEER_GROUP_HEADER:
+                state = PEER_GROUPS
+            elif  state == PEER_GROUPS:
+                if regex_prefix_peers.match(line):
+                    state = PREFIX_PATHS
+                else:
+                    state = ERR
+            elif state == PREFIX_PATHS:
+                if regex_prefix_paths.match(line):
+                    path = line.strip()
+                    self.facts['bgp_route'][prefix]['paths'].append(path)
+                    state = PREFIX_PATHS_FROM
+                else:
+                    state = ERR
+            elif state == PREFIX_PATHS_FROM:
+                if regex_prefix_path_p1_from.match(line):
+                    state = PREFIX_PATH_ORIGIN
+                else:
+                    state = ERR
+            elif state == PREFIX_PATH_ORIGIN:
+                if regex_prefix_path_p2_origin.match(line):
+                    state = PREFIX_PATH_TIMESTAMP
+                else:
+                    state = ERR
+            elif state == PREFIX_PATH_TIMESTAMP:
+                if regex_prefix_path_p3_timestamp.match(line):
+                    state = PREFIX_PATHS
+                else:
+                    state = ERR
+            elif state == ERR:
+                raise Exception("cannot parse bgp prefix info correctly " + str(state) + str(self.facts))
+
+
+def collect_data(module, command=None):
+    """
+       Collect bgp information by reading output of 'vtysh' command line tool
+    """
+    rc, out, err = module.run_command('vtysh -c "show ip bgp ' + command + '"',
+                                        executable='/bin/bash', use_unsafe_shell=True)
+    return (rc, out, err)
+
+
+def main():
+    module = AnsibleModule(
+            argument_spec=dict(
+                neighbor=dict(required=False, default=None),
+                direction=dict(required=False, choices=['adv', 'rec']),
+                prefix=dict(required=False, default=None)
+                ),
+            supports_check_mode=False
+            )
+    m_args = module.params
+    neigh = m_args['neighbor']
+    direction = m_args['direction']
+    prefix = m_args['prefix']
```

Comment:
overall, we should first to parameter validation and raise any exception is parameters are incorrect.

Then, we run the command and get the results, raise errors when the return results are incorrect.

we should not mixed them in one block.